In [16]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler

# load dataset
df = pd.read_csv('/Users/urvashibalasubramaniam/Documents/GitHub/Kabir_Vohra_Urvashi_Balasubraniam_A1/retail_task/src/Retail.csv')

ParserError: Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.

First, understanding the dataset by checking its (1) categories (2) stats. Then check for any missing fields. 

In [ ]:
# categories
df.columns

Index(['age', 'gender', 'income_bracket', 'loyalty_program',
       'membership_years', 'churned', 'marital_status', 'number_of_children',
       'education_level', 'occupation', 'transaction_id', 'transaction_date',
       'product_id', 'product_category', 'quantity', 'unit_price',
       'discount_applied', 'payment_method', 'store_location',
       'transaction_hour', 'day_of_week', 'week_of_year', 'month_of_year',
       'avg_purchase_value', 'purchase_frequency', 'last_purchase_date',
       'avg_discount_used', 'preferred_store', 'online_purchases',
       'in_store_purchases', 'avg_items_per_transaction',
       'avg_transaction_value', 'total_returned_items', 'total_returned_value',
       'total_sales', 'total_transactions', 'total_items_purchased',
       'total_discounts_received', 'avg_spent_per_category',
       'max_single_purchase_value', 'min_single_purchase_value',
       'product_name', 'product_brand', 'product_rating',
       'product_review_count', 'product_stock',

In [ ]:
# describe in common statistical terms
df.describe()

age  membership_years  number_of_children  transaction_id  \
count  800000.000000     800000.000000       800000.000000   800000.000000   
mean       48.489880          4.498445            1.999791   499796.925741   
std        17.868101          2.872934            1.414198   288558.279459   
min        18.000000          0.000000            0.000000        2.000000   
25%        33.000000          2.000000            1.000000   249944.000000   
50%        49.000000          4.000000            2.000000   499641.500000   
75%        64.000000          7.000000            3.000000   749525.250000   
max        79.000000          9.000000            4.000000   999999.000000   

          product_id       quantity     unit_price  discount_applied  \
count  800000.000000  800000.000000  800000.000000     800000.000000   
mean     4999.302701       5.003425     500.120805          0.250013   
std      2887.681227       2.583392     288.440234          0.144315   
min         1.000000       1.000000       1.000000          0.000000   
25%      2495.000000       3.000000     250.130000          0.130000   
50%      4999.000000       5.000000     500.120000          0.250000   
75%      7501.000000       7.000000     749.940000          0.370000   
max      9999.000000       9.000000    1000.000000          0.500000   

       transaction_hour   week_of_year  ...  product_return_rate  \
count     800000.000000  800000.000000  ...        800000.000000   
mean          11.508469      26.501809  ...             0.250098   
std            6.925024      15.005291  ...             0.144413   
min            0.000000       1.000000  ...             0.000000   
25%            5.000000      14.000000  ...             0.130000   
50%           12.000000      27.000000  ...             0.250000   
75%           18.000000      39.000000  ...             0.380000   
max           23.000000      52.000000  ...             0.500000   

       product_weight  product_shelf_life   promotion_id  customer_zip_code  \
count   800000.000000       800000.000000  800000.000000      800000.000000   
mean         5.056466          181.891146     499.726743       55002.041589   
std          2.857353          105.272243     288.542749       25981.823465   
min          0.100000            0.000000       1.000000       10000.000000   
25%          2.580000           91.000000     250.000000       32464.000000   
50%          5.060000          182.000000     500.000000       54997.000000   
75%          7.530000          273.000000     750.000000       77507.000000   
max         10.000000          364.000000     999.000000       99998.000000   

       store_zip_code  distance_to_store  customer_support_calls  \
count   800000.000000      800000.000000           800000.000000   
mean     54963.376260          50.011425                9.498734   
std      25975.911014          28.863479                5.761851   
min      10000.000000           0.000000                0.000000   
25%      32484.000000          25.000000                4.000000   
50%      54933.000000          50.010000                9.000000   
75%      77442.000000          74.990000               14.000000   
max      99998.000000         100.000000               19.000000   

       website_visits  days_since_last_purchase  
count   800000.000000             800000.000000  
mean        49.511601                182.076024  
std         28.864197                105.348022  
min          0.000000                  0.000000  
25%         25.000000                 91.000000  
50%         50.000000                182.000000  
75%         75.000000                273.000000  
max         99.000000                364.000000  

[8 rows x 39 columns]

In [ ]:
# get categories (header values)
print(df.iloc[0])

age                             41
gender                      Female
income_bracket              Medium
loyalty_program                 No
membership_years                 3
                             ...  
email_subscriptions             No
app_usage                   Medium
website_visits                   2
social_media_engagement     Medium
days_since_last_purchase         6
Name: 0, Length: 77, dtype: object


# Check for missing values

In [ ]:
df.info() # will give us count of non-null entries, among other things

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800000 entries, 0 to 799999
Data columns (total 77 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   age                        800000 non-null  int64  
 1   gender                     800000 non-null  object 
 2   income_bracket             800000 non-null  object 
 3   loyalty_program            800000 non-null  object 
 4   membership_years           800000 non-null  int64  
 5   churned                    800000 non-null  object 
 6   marital_status             800000 non-null  object 
 7   number_of_children         800000 non-null  int64  
 8   education_level            800000 non-null  object 
 9   occupation                 800000 non-null  object 
 10  transaction_id             800000 non-null  int64  
 11  transaction_date           800000 non-null  object 
 12  product_id                 800000 non-null  int64  
 13  product_category           80

All columns have n (80,000) non-null values => no missing fields!

# Correlation Analysis
Numerical-numerical columns tested with corr()
Numerical-categorical columns tested with Point Biserial Correlation (after one-hot encoding categorical data)

In [ ]:
# correlation analysis -- every variable with target
# Separating numerical and categorical (non-numerical/string) data value columns
numericalCols = df.select_dtypes(include=["number"]).columns.tolist()

# wrt avg_purchase_value (target feature)
correlations = df[numericalCols].corr()["avg_purchase_value"]

# sort correlation values
correlations_sorted = correlations.sort_values(ascending=False)

# extract top 10 and bottom 10
print(correlations_sorted.head(10), "\n")
print("bottom:")
print(correlations_sorted.tail(10), "\n")


avg_purchase_value          1.000000
total_returned_items        0.002510
in_store_purchases          0.002482
quantity                    0.002272
transaction_id              0.001726
total_items_purchased       0.001661
website_visits              0.001527
customer_zip_code           0.001292
days_since_last_purchase    0.000979
product_review_count        0.000955
Name: avg_purchase_value, dtype: float64 

bottom:
number_of_children          -0.001362
min_single_purchase_value   -0.001418
membership_years            -0.001789
product_return_rate         -0.002060
online_purchases            -0.002129
distance_to_store           -0.002428
product_rating              -0.002480
avg_discount_used           -0.002486
total_transactions          -0.002650
customer_support_calls      -0.002653
Name: avg_purchase_value, dtype: float64 



Checking how many unique values exist per categorical column and thus checking if one-hot encoding is feasible:

In [ ]:
# select categorical cols
categoryCols = df.select_dtypes(exclude=["number"]).columns.tolist()

# loop through categorical variables, count unique values
print("Column name : number of unique values")
print("--------------------------------------")
for col_name in categoryCols:
    print(f"'{col_name}' : {df[col_name].nunique()}") # nunique counts unique vals

Column name : number of unique values
--------------------------------------
'gender' : 3
'income_bracket' : 3
'loyalty_program' : 2
'churned' : 2
'marital_status' : 3
'education_level' : 4
'occupation' : 4
'transaction_date' : 795027
'product_category' : 5
'payment_method' : 4
'store_location' : 4
'day_of_week' : 7
'purchase_frequency' : 4
'last_purchase_date' : 789895
'preferred_store' : 4
'product_name' : 4
'product_brand' : 3
'product_size' : 3
'product_color' : 5
'product_material' : 4
'product_manufacture_date' : 794914
'product_expiry_date' : 794875
'promotion_type' : 3
'promotion_start_date' : 789928
'promotion_end_date' : 789906
'promotion_effectiveness' : 3
'promotion_channel' : 3
'promotion_target_audience' : 2
'customer_city' : 4
'customer_state' : 3
'store_city' : 4
'store_state' : 3
'holiday_season' : 2
'season' : 4
'weekend' : 2
'email_subscriptions' : 2
'app_usage' : 3
'social_media_engagement' : 3


One hot encoding is feasible for all non-date values. 
Extracting data from relevant date values below:

# Date Features Analysis: 
What data could I potentially extract from the date columns?

*_'transaction_date' : 795027*_
Seasonality (extract quarter)
Week of the year – already a separate feature in my dataset
Month of the year – already given in “month of year”
Time of day – already given in “transaction hour”
Weekend indicator (yes/no binary value) – already done in “day of the week”

Therefore, delete this column as it provides no new information.

*_'last_purchase_date' : 789895*_
We already have purchase frequency
Could convert to “days since last purchase” – but this would have to be dynamically updated according to the current date, and still doesn’t provide much new information.

Therefore, delete this column as it provides little new information.

*_'product_manufacture_date' : 794914*_
*_'product_expiry_date' : 794875*_
This interval is already represented clearly in product_shelf_life

Therefore, delete this column as it provides no new information.

*_'promotion_start_date' : 789928*_
*_'promotion_end_date' : 789906*_

Convert this column to an interval (*promotion_time_period*) that calculates the number of days between start and end date

Creating interval column for promotion_time_period
and deleting all other date columns in place:

In [ ]:
# converting numerical to datetime datatype (for interval calculation)
df['promotion_start_date'] = pd.to_datetime(df['promotion_start_date'])
df['promotion_end_date'] = pd.to_datetime(df['promotion_end_date'])

# verifying if they are both successfully converted (then output=datetime64ns)
print(df['promotion_start_date'].dtype) 
print(df['promotion_end_date'].dtype) 

# creating new column in df = interval b/w promotion start and end date (unit = days)
df['promotion_time_period'] = (df['promotion_end_date'] - df['promotion_start_date']).dt.days

# dropping (deleting) date columns in new dataframe
df_new = df.drop(['transaction_date','last_purchase_date','product_manufacture_date','product_expiry_date','promotion_start_date','promotion_end_date'], axis=1)

datetime64[ns]
datetime64[ns]


In [ ]:
df_new.columns # check if the operation worked correctly

Index(['age', 'gender', 'income_bracket', 'loyalty_program',
       'membership_years', 'churned', 'marital_status', 'number_of_children',
       'education_level', 'occupation', 'transaction_id', 'product_id',
       'product_category', 'quantity', 'unit_price', 'discount_applied',
       'payment_method', 'store_location', 'transaction_hour', 'day_of_week',
       'week_of_year', 'month_of_year', 'avg_purchase_value',
       'purchase_frequency', 'avg_discount_used', 'preferred_store',
       'online_purchases', 'in_store_purchases', 'avg_items_per_transaction',
       'avg_transaction_value', 'total_returned_items', 'total_returned_value',
       'total_sales', 'total_transactions', 'total_items_purchased',
       'total_discounts_received', 'avg_spent_per_category',
       'max_single_purchase_value', 'min_single_purchase_value',
       'product_name', 'product_brand', 'product_rating',
       'product_review_count', 'product_stock', 'product_return_rate',
       'product_size', 'p

In [ ]:
# overwrite df with df_new
df = df_new

# One Hot Encoding
Now that the date columns are cleaned up, performing one-hot encoding on remaining variables (all of which have n<10 columns)

In [ ]:
# getting category cols, from overwritten (just modified) df
categoryCols = df.select_dtypes(exclude=["number"]).columns.tolist()

df_encoded = pd.get_dummies(df, columns=categoryCols)

In [ ]:
print(df_encoded.head())

   age  membership_years  number_of_children  transaction_id  product_id  \
0   41                 3                   0          237855         997   
1   40                 4                   4          500661        2869   
2   48                 4                   3          432552        6172   
3   19                 9                   3          771343        6004   
4   53                 8                   2          186219        3565   

   quantity  unit_price  discount_applied  transaction_hour  week_of_year  \
0         7      960.62              0.24                 0            48   
1         5      219.52              0.46                12            19   
2         8      759.95              0.11                22             1   
3         9      862.15              0.44                11            36   
4         3      609.81              0.19                11            42   

   ...  weekend_No  weekend_Yes  email_subscriptions_No  \
0  ...        True   

In [17]:
df = df_encoded # overwriting as it is correctly doing the operation

# typecasting all bools to integers (False -> 0, True -> 1)
boolean_cols = df.select_dtypes(include=['bool']).columns
df[boolean_cols] = df[boolean_cols].astype(int) 

print(df.head())

   age  membership_years  number_of_children  transaction_id  product_id  \
0   41                 3                   0          237855         997   
1   40                 4                   4          500661        2869   
2   48                 4                   3          432552        6172   
3   19                 9                   3          771343        6004   
4   53                 8                   2          186219        3565   

   quantity  unit_price  discount_applied  transaction_hour  week_of_year  \
0         7      960.62              0.24                 0            48   
1         5      219.52              0.46                12            19   
2         8      759.95              0.11                22             1   
3         9      862.15              0.44                11            36   
4         3      609.81              0.19                11            42   

   ...  weekend_No  weekend_Yes  email_subscriptions_No  \
0  ...           1   

Testing correlation for categorical variables with target (numerial) feature, using the Point Biserial test: